src/blog/2024/02/postgresql-with-node-red.md

```diff
@@ -0,0 +1,218 @@
+
+1. Install `node-red-contrib-postgresql` by the pallet manager, you can choose other nodes according to your preference, but we chose this node because it is part of a flowfuse-certified node which assures me that the node is robust, secure, and developed with high-quality. For more details refer to our [certified node](https://flowfuse.com/blog/2023/10/certified-nodes/) blog post.
+2. Before connecting to your PostgreSQL database, ensure you have the following information ready. Later We will add environment variables to configure the PostgreSQL node to connect with the database:
+
+- Host: IP address or hostname of your PostgreSQL server.
+- Port: By default, PostgreSQL uses port 5432. Ensure this matches your PostgreSQL server configuration.
+- Database: The name of the PostgreSQL database you want to connect to.
+- User: username with the necessary privileges to access the specified database.
+- Password: corresponding password for the username.
+
+# Adding environment variables
+
+We have discussed many times in previous blogs that using environment variables is a best practice that prevents revealing confidential configuration details such as API keys, passwords, secret keys, etc directly in the flow. Therefore, In this section, we will add environment variables for Postgresql configuration details, for more details see  [Using Environment Variables in Node-RED](https://flowfuse.com/blog/2023/02/environment-variables-in-node-red/).
+
+1. Navigate to the instance's setting and then go to the environment section.
+2. Click on the `add variable` button and add variables for each of the configuration data that we discussed in the above section.
+3. Click on the save button and restart the instance by clicking on the top right `Action` button and selecting the restart option.
+
+!["Adding environment variables"](./images/postgresql_with_node-red_environment_variables.png "Adding environment variables")
+
+# Creating Table
+
+In this section, we will create a table in our database to store product data. Additionally, it's important to note that throughout this guide we will send an object containing the properties of configuration details with each request to the database. This approach allows us to use environment variables to secure our sensitive configuration details while connecting to the database.
+
+!["Setting pgConfig objects property"](./images/postgresql_with_node-red_change_node.png "Setting pgConfig objects property")
+
+1. Drag an Inject node onto canvas, and keep it unchanged.
+2. Drag the change node onto canvas, and set `msg.pgConfig` properties using the added environment variable as shown in the above image.
+3. Drag a Postgresql node onto Canvas, click on that node, and paste the following SQL command into the query input field (I have added comments for your understanding of sql commands)
```

Comment:
@sumitshinde-84 Why aren't you using the PG config node to setup these details?

Reply:
That package has only one node, which we are currently using. Additionally, configuration via the nodes tab doesn't support environment variables.

Reply:
> configuration via the nodes tab doesn't support environment variables.

Do you not have the $ selection drop down in your PG config?

![image](https://github.com/FlowFuse/website/assets/44235289/3bb3d2ba-08ae-493f-9ad0-1630becdff36)

![image](https://github.com/FlowFuse/website/assets/44235289/a2b55edb-ab0c-40ad-88cc-cca586460dab)

![image](https://github.com/FlowFuse/website/assets/44235289/741b9435-2b7f-4040-a8ab-08501d56cddb)

Reply:
@Steve-Mcl, configuring the pg config node only supports Flow level environment variables or environment variables set within the settings -> environment, but it doesn't support global environment variables.

Reply:
That is quite odd if true. The expectation of `env` setting on a `typedInput` control should not be limited to any scope at all.

There is always the fall back option of using `${PG_HOST}` etc. That cannot be _incorrectly_ interpreted by a node author since it is done at start up time of Node-RED (the `${PG_HOST}` value gets replaced before the contrib node knows about it)

So, in short, if the `env` setting does not work, using `${xxx}` will work as is described in the Node-RED docs: https://nodered.org/docs/user-guide/environment-variables

![image](https://github.com/FlowFuse/website/assets/44235289/fa2e6cab-5f2f-4cea-8bcf-2f476e5be6ad)

Reply:
@Steve-Mcl  could you please take look at issue which I have created at authors repo - https://github.com/alexandrainst/node-red-contrib-postgresql/issues/64

Reply:
> As to why the config node is a better option, as you can see in the source code, a `new client` is created every time you supply the `pgConfig` in the `msg`. With a config, the pool is used to save generating new connections to the DB
> 
> REF: https://github.com/alexandrainst/node-red-contrib-postgresql/blob/main/postgresql.js#L222-L227
> 
> 
> > but it doesn't support global environment variables.
> 
> I am struggling to see how that is the case. Looking at the code, if you set `env` type, the code gets it from `process.env` (which is where env vars are presented to the process) https://github.com/alexandrainst/node-red-contrib-postgresql/blob/93a4f0dacb1f6b2214778356cece7516ec68b58b/postgresql.js#L48
> 
> Perhaps I missing some finer detail? Are you actually referring to Node-RED configured Global Env Vars (not the Env Vars FlowFuse provides to the instance)?

Okay @Steve-Mcl, I will update the configuration section in the post to use the pgconfig node.

Reply:
Thanks @Steve-Mcl for providing a nice demo and information. Furthermore, regarding the post, I will update it to configure PostgreSQL via the pgConfig tab/node directly. Creating an instance of the client every time is indeed not the best practice

Reply:
Excellent!

Happy to help @sumitshinde-84 

Like you, I'd rather best practice is encouraged.